processor/src/main/java/org/mapstruct/ap/internal/util/workarounds/SpecificCompilerWorkarounds.java

```diff
@@ -149,4 +156,45 @@ static TypeMirror asMemberOf(Types typeUtils, ProcessingEnvironment env, Version
 
         return result;
     }
+
+    /**
+     * In Eclipse 4.6, the ClassLoader of the annotation processor does not provide access to the implementation types
+     * of the APT-API anymore, so we need to create a new ClassLoader containing the processor class path URLs and
+     * having the ClassLoader of the APT-API implementations as parent. The method invocation then consequently needs to
+     * be done via reflection.
+     *
+     * @author Andreas Gudian
+     */
+    private static final class ClassLoaderHelper {
+        private static ClassLoader classLoader;
+        private static Method asMemberOf;
+
+        private static ClassLoader getOrCreateClassLoader(ClassLoader parent) {
+            if ( classLoader == null ) {
+                classLoader = new URLClassLoader(
+                    ( (URLClassLoader) ClassLoaderHelper.class.getClassLoader() ).getURLs(), parent );
+            }
+
+            return classLoader;
+        }
+
+        private static Method getAsMemberOf(ClassLoader platformClassLoader) throws Exception {
+            if ( asMemberOf == null ) {
+                Class<?> workaroundClass = getOrCreateClassLoader( platformClassLoader )
+                    .loadClass( ECLIPSE_AS_MEMBER_OF_WORKAROUND );
```

Comment:
Because our class itself and the classes it uses (ElementImpl and so on) need to be loaded from the same classloader hierarchy. As such a classloader doesn't exist, I need to create a new one - with the compiler's classloader as parent as I need to interact with those types directly.
With our workaround-class being loaded in that nwe classloader that has no relation to the CL of the rest of MapStruct, I can't cast that newly loaded class here to one of our types. Hence the need to do it via reflection.
The alternative would be to try and do everything that's being done in our workaround class on the eclipse specific types solely via reflection. But that's no fun either ;-).